nolearn/lasagne/handlers.py

```diff
@@ -92,3 +91,78 @@ def __call__(self, nn, train_history):
                 pickle.dump(nn, f, -1)
         else:
             nn.save_params_to(path)
+
+
+class PrintLayerInfo:
+    def __init__(self):
+        pass
+
+    def __call__(self, nn, train_history):
+        message = self._get_greeting(nn)
+        print(message)
+        print("## Layer information")
+        print("")
+
+        layers_contain_conv2d = is_conv2d(nn.layers_.values())
+        if not layers_contain_conv2d or (nn.verbose < 2):
+            layer_info = self._get_layer_info_plain(nn)
+            legend = None
+        else:
+            layer_info, legend = self._get_layer_info_conv(nn)
+        print(layer_info)
+        if legend is not None:
+            print(legend)
+        print("")
+
+    @staticmethod
+    def _get_greeting(nn):
+        shapes = [param.get_value().shape for param in
+                  nn.get_all_params() if param]
+        nparams = reduce(operator.add, [reduce(operator.mul, shape) for
+                                        shape in shapes])
+        message = ("# Neural Network with {} learnable parameters"
+                   "\n".format(nparams))
+        return message
+
+    @staticmethod
+    def _get_layer_info_plain(nn):
+        nums = list(range(len(nn.layers)))
+        names = [layer.name for layer in nn.layers_.values()]
+        output_shapes = ['x'.join(map(str, layer.get_output_shape()[1:]))
+                         for layer in nn.layers_.values()]
+
+        table = OrderedDict([
+            ('#', nums),
+            ('name', names),
+            ('size', output_shapes),
+        ])
+        layer_infos = tabulate(table, 'keys', tablefmt='pipe')
+        return layer_infos
+
+    @staticmethod
+    def _get_layer_info_conv(nn):
+        if nn.verbose > 2:
+            detailed = True
+            tablefmt = 'simple'
+        else:
+            detailed = False
+            tablefmt = 'pipe'
```

Comment:
Can't we just use the default tablefmt everywhere?  What's the deal with `pipe`, why is it preferred sometimes?